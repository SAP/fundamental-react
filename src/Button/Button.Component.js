import path from 'path';
import React from 'react';
import { Button, ButtonGroup } from '../';
import { ComponentPage, Example } from '../_playground';

export const ButtonComponent = () => {
    const clickBtnHandler = btn => {
        alert(`You clicked the ${btn} Button`);
    };

    return (
        <ComponentPage
            description={`A **Button** allows users to perform an action. The priority of buttons within a page should be considered.
                For instance, only use the main button once within a page or dialog. Color is also important. For
                instance, the most important button has a blue background where as a red button should only be used if
                the action it performs is potentially destructive.`}
            sourceModulePath={path.join(__dirname, './Button')}
            title='Button'>

            <Example
                centered
                description={`There are three emphasis styles used to indicate the importance of the button on
                    the page.\n\n* **Emphasized Button**: There should only be one highlighted button on the page.
                    This is the primary call to action.\n\n* **Regular Button**: The default button style and the
                    most common button. There may be more than one on a page.\n\n* **Transparent Button**: This is the
                    lowest priority button and most often used with page content like appearing in a table or list.
                    There may be more than one on the page.`}
                title='Button Options'>
                <Button onClick={() => clickBtnHandler('Emphasized')} option='emphasized'>
                        Emphasized Button
                </Button>
                <Button onClick={() => clickBtnHandler('Regular')}>Regular Button</Button>
                <Button onClick={() => clickBtnHandler('Transparent')} option='transparent'>Transparent Button</Button>
            </Example>

            <Example
                centered
                description={`* **Action Button**: The default button \n\n* **Standard Button**:
                    Neutral or informative color \n\n* **Positive Button**: Used for positive actions 
                    such as approved, ok, yes. \n\n* **Negative Button**: Used for negative actions such as decline, cancel, no.`}
                title='Button Types'>
                <Button>Action Button</Button>
                <Button type='standard'>Standard Button</Button>
                <Button type='positive'>Positive Button</Button>
                <Button type='negative'>Negative Button</Button>
            </Example>

            <Example
                centered
                description={`Button can have an icon with text or just and icon. You can use \`glyph="icon-name"\` to
                    attach an icon to the button.`}
                title='Buttons with Icon'>
                <div className='fd-container'>
                    <div className='fd-col--6'>
                        <Button glyph='cart' option='emphasized'>
                                Add to Cart
                        </Button>

                        <Button glyph='cart'>Add to Cart</Button>

                        <Button glyph='filter' option='transparent'>
                                Add to Cart
                        </Button>

                        <Button glyph='accept' type='positive'>
                                Approve
                        </Button>

                        <Button glyph='decline' type='negative'>
                                Reject
                        </Button>

                        <Button
                            glyph='alert'
                            option='emphasized'>
                                Review
                        </Button>
                    </div>

                    <div className='fd-col--6'>
                        <Button glyph='cart' />

                        <Button glyph='cart' option='transparent' />

                        <Button glyph='filter' type='standard' />

                        <Button glyph='accept' type='positive' />

                        <Button glyph='decline' type='negative' />
                    </div>
                </div>
            </Example>

            <Example
                centered
                description={`There are two sizes. The \`compact\` size is only used on desktop and it is full size when used
                    on a touch device.`}
                title='Button Sizes'>
                <Button>Default</Button>
                <Button compact>Compact</Button>
            </Example>

            <Example
                centered
                description={`There are three states: default, \`selected\`,
                    and \`disabled\`.`}
                title='Button States'>
                <div className='frDocs-tile__centered'>
                    <Button>Default</Button>
                    <Button selected>Selected</Button>
                    <Button disabled>Disabled</Button>
                </div>
                <div className='frDocs-tile__break' />
                <div className='frDocs-tile__centered'>
                    <Button option='emphasized'>Emphasized</Button>
                    <Button option='emphasized' selected>
                            Selected
                    </Button>
                    <Button disabled option='emphasized'>
                            Disabled
                    </Button>
                </div>
                <div className='frDocs-tile__break' />
                <div className='frDocs-tile__centered'>
                    <Button option='transparent'>Transparent</Button>
                    <Button option='transparent' selected>
                    Selected
                    </Button>
                    <Button disabled option='transparent'>
                    Disabled
                    </Button>
                </div>
                <div className='frDocs-tile__break' />
                <div className='frDocs-tile__centered'>
                    <Button type='positive'>Positive</Button>
                    <Button selected type='positive'>
                    Selected
                    </Button>
                    <Button disabled type='positive'>
                    Disabled
                    </Button>
                </div>
                <div className='frDocs-tile__break' />
                <div className='frDocs-tile__centered'>
                    <Button type='negative'>Negative</Button>
                    <Button selected type='negative'>
                    Selected
                    </Button>
                    <Button disabled type='negative'>
                    Disabled
                    </Button>
                </div>
            </Example>

            <Example
                centered
                title='Button Group'>
                <ButtonGroup>
                    <Button glyph='survey' />
                    <Button glyph='pie-chart' selected />
                    <Button glyph='pool' />
                </ButtonGroup>

                <ButtonGroup>
                    <Button compact>Left</Button>
                    <Button compact selected>
                            Middle
                    </Button>
                    <Button compact>Right</Button>
                </ButtonGroup>
            </Example>
        </ComponentPage>
    );
};
