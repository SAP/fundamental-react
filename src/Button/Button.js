import classnamesBind from 'classnames/bind';
import Icon from '../Icon/Icon';
import PropTypes from 'prop-types';
import React from 'react';
import useUniqueId from '../utils/useUniqueId';
import withStyles from '../utils/withStyles';
import { BUTTON_OPTIONS, BUTTON_TYPES } from '../utils/constants';
import styles from 'fundamental-styles/dist/button.css';

const classnames = classnamesBind.bind(styles);

/** A **Button** allows users to perform an action. The priority of buttons within a page should be considered.
For instance, only use the main button once within a page or dialog. Color is also important. For
instance, the most important button has a blue background where as a red button should only be used if
the action it performs is potentially destructive. */

const Button = React.forwardRef(({
    allowFocusOnDisable,
    option,
    type,
    compact,
    disabledMessage,
    enabledMessage,
    glyph,
    iconBeforeText,
    iconProps,
    selected,
    disabled,
    typeAttr,
    onClick,
    children,
    className,
    cssNamespace,
    textClassName,
    badge,
    badgeClassName,
    ...props
}, ref) => {

    const buttonClasses = classnames(
        `${cssNamespace}-button`,
        {
            [`${cssNamespace}-button--${option}`]: !!option,
            [`${cssNamespace}-button--${type}`]: !!type,
            [`${cssNamespace}-button--compact`]: compact,
            'is-selected': selected,
            'is-disabled': disabled
        },
        className
    );

    const buttonTextClasses = classnames(
        `${cssNamespace}-button__text`,
        textClassName
    );

    const badgeClasses = classnames(
        `${cssNamespace}-button__badge`,
        badgeClassName
    );

    const ariaDisabled = props['aria-disabled'];
    let disabledInAnyWay = disabled || ariaDisabled;

    //if button is disabled and focusable, remove onClick handler
    let onClickHandler = allowFocusOnDisable && disabledInAnyWay ? () => {} : onClick;

    const ariaLiveId = useUniqueId();

    const getDisabledProps = () => {
        let disabledProps = {};
        if (allowFocusOnDisable) {
            disabledProps['aria-disabled'] = disabledInAnyWay;
            disabledProps['aria-describedby'] = ariaLiveId;
        } else {
            disabledProps.disabled = disabledInAnyWay;
        }
        return disabledProps;
    };

    const liveMessageForScreenReaders = disabledInAnyWay ? disabledMessage : enabledMessage;

    const renderButtonStateMessage = () => {
        let content = null;
        if (allowFocusOnDisable && disabledInAnyWay !== null) {
            content = (
                <p
                    aria-live='assertive'
                    className={classnames(`${cssNamespace}-button__instructions`)}
                    id={ariaLiveId}>
                    {liveMessageForScreenReaders}
                </p>)
            ;
        }
        return content;
    };

    const renderIcon = () => {
        const content = glyph ? (
            <Icon
                {...iconProps}
                ariaHidden
                glyph={glyph} />
        ) : null;

        return content;
    };

    return (
        <>
            <button
                {...props}
                {...getDisabledProps()}
                className={buttonClasses}
                onClick={onClickHandler}
                ref={ref}
                type={typeAttr}>
                {iconBeforeText && renderIcon()}
                {children && <span className={buttonTextClasses}>{children}</span>}
                {!iconBeforeText && renderIcon()}
                {badge && <span className={badgeClasses}>{badge}</span>}
            </button>
            {renderButtonStateMessage()}
        </>
    );
});

Button.displayName = 'Button';

const validateStateTransitionMessage = (props, propName) => {
    if (props.allowFocusOnDisable && (props.disabled !== null || props['aria-disabled']) && !props[propName]) {
        return new Error(`
        Missing or invalid value for ${propName}.
        If this button can receive focus when disabled (i.e. allowFocusOnDisable prop is true),
        you need to provide localized string values for the enabled and disabled states.
        These strings are used to announce the state change by assistive technologies.
        Hence, these should be specified for a better, accessible experience.
        For e.g. 
        <Button
            disabled={someLogic()}
            allowFocusOnDisable
            disabledMessage="Please complete required form fields to enable form submission."
            enabledMessage="Please press the submit button to proceed.">
            Submit
        </Button>
        `);
    }
};

const badgePropValidation = (props, propName) => {
    if (props[propName]?.length < 1 || props[propName]?.length > 4) {
        return new Error(`${propName} must be between 1-4 characters.`);
    }
};

Button.propTypes = {
    /** Set to **true** to enable button focus through keyboard on a disabled button */
    allowFocusOnDisable: PropTypes.bool,
    /** Automatically set when disabled prop is passed into button */
    'aria-disabled': PropTypes.bool,
    /** Content to display in a badge. Must be between 1-4 characters */
    badge: badgePropValidation,
    /** CSS class(es) to add to the badge element */
    badgeClassName: PropTypes.string,
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to enable compact mode */
    compact: PropTypes.bool,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Message for screen reader to inform user the button is disabled */
    disabledMessage: validateStateTransitionMessage,
    /** Message for screen reader to inform user the button is enabled */
    enabledMessage: validateStateTransitionMessage,
    /** The icon to include. See the icon page for the list of icons */
    glyph: PropTypes.string,
    /** Determines whether the icon should be placed before the text */
    iconBeforeText: PropTypes.bool,
    /** Additional props to be spread to the Icon component */
    iconProps: PropTypes.object,
    /** Indicates the importance of the button: 'emphasized' or 'transparent' */
    option: PropTypes.oneOf(BUTTON_OPTIONS),
    /** Set to **true** to set state of the button to "selected" */
    selected: PropTypes.bool,
    /** CSS class(es) to add to the text element */
    textClassName: PropTypes.string,
    /** Sets the variation of the component. Primarily used for styling: 'standard',
    'positive',
    'negative',
    'medium' */
    type: PropTypes.oneOf(BUTTON_TYPES),
    /** Value for the `type` attribute on the `<button>` element */
    typeAttr: PropTypes.string,
    /**
     * Callback function; triggered when the button is clicked.
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent. See https://reactjs.org/docs/events.html.
     * @returns {void}
    */
    onClick: PropTypes.func
};

Button.defaultProps = {
    typeAttr: 'button'
};

export default withStyles(Button);
