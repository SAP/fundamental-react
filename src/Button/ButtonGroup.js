import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/WithStyles/WithStyles';

const ButtonGroup = React.forwardRef(({ children, disableStyles, ...props }, ref) => {
    return (
        <div
            {...props}
            aria-label='Group label'
            className='fd-button-group'
            ref={ref}
            role='group'>
            {children}
        </div>
    );
});

ButtonGroup.displayName = 'ButtonGroup';

ButtonGroup.propTypes = {
    children: PropTypes.node,
    customStyles: PropTypes.object,
    disableStyles: PropTypes.bool
};

ButtonGroup.propDescriptions = {
    children: '`Button` nodes to group together.'
};

export { ButtonGroup as __ButtonGroup };

export default withStyles(ButtonGroup, { cssFile: 'button-group' });
