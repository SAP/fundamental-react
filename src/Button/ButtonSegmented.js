import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/segmented-button.css';

const classnames = classnamesBind.bind(styles);

/** Group a series of buttons together on a single line with the segmented button.
 * Only one of the options can be active at a time, the others remain or become inactive.
 * The option can be activated by clicking on it.
 * This type of button is comparable to a radio button group. */
const ButtonSegmented = React.forwardRef(({ children, cssNamespace, ...props }, ref) => (
    <div
        {...props}
        aria-label='Group label'
        className={classnames(`${cssNamespace}-segmented-button`)}
        ref={ref}
        role='group'>
        {children}
    </div>)
);

ButtonSegmented.displayName = 'ButtonSegmented';

ButtonSegmented.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node
};

export default withStyles(ButtonSegmented);
