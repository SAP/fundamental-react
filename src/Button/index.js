export { default as Button } from './Button.js';
export { default as ButtonGroup } from './ButtonGroup.js';
