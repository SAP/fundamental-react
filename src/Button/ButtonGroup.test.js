import Button from './Button';
import ButtonGroup from './ButtonGroup';
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<ButtonGroup />', () => {
    const buttonGroup = (
        <ButtonGroup>
            <Button className='blue' glyph='survey' />
            <Button glyph='pie-chart' selected />
            <Button glyph='pool' />
        </ButtonGroup>
    );

    test('create button group', () => {
        const component = renderer.create(buttonGroup);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the ButtonGroup component', () => {
            const element = mount(<ButtonGroup data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });
    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <ButtonGroup ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('DIV');
        expect(ref.current.className).toEqual('fd-button-group');
    });
});
