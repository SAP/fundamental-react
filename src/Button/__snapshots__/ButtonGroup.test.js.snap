// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> create button group 1`] = `
<div
  aria-label="Group label"
  className="fd-button-group"
  role="group"
>
  <button
    className="fd-button sap-icon--survey blue"
  />
  <button
    className="fd-button sap-icon--pie-chart is-selected"
    selected={true}
  />
  <button
    className="fd-button sap-icon--pool"
  />
</div>
`;
